File: p210m/course_schedule2.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function(numCourses, prerequisites) {
    // Convert the input edges into a graph
    const graph = (function makeGraph() {
        let graph = Array.from({length: numCourses}, v => []);
        for (let e of prerequisites) {
            graph[e[0]].push(e[1]);
        }
        return graph;
    })();
    // console.log(graph);
    /**
     * Conduct topological sort and check cycle
     * @return true if having cycle 
     */
    const recTopologicalSort = function(v, visited, onPath, sorted) {
        if (visited[v]) { return false; }
        visited[v] = true; onPath[v] = true;
        for (let i of graph[v]) {
            if (onPath[i] || recTopologicalSort(i, visited, onPath, sorted)) {
                return true;
            }
        }
        onPath[v] = false;
        sorted.push(v);
        return false;
    };
    let sorted = [];
    let visited = Array.from({length: numCourses}, v => false);
    let onPath = Array.from(visited);
    for (let v = 0; v < numCourses; ++v) {
        if (!visited[v] && recTopologicalSort(v, visited, onPath, sorted)) {
            return [];
        }
    }
    return sorted;
};
// TEST
// TEST
[
    {
        numCourses: 1,
        prerequisites: []
    },
    {
        numCourses: 2,
        prerequisites: [[0,1]]
    },
    {
        numCourses: 2,
        prerequisites: [[0,1],[1,0]]
    },
    {
        numCourses: 4,
        prerequisites: [[0,1],[1,2],[3,1]]
    },
    {
        numCourses: 4,
        prerequisites: [[0,1],[1,2],[2,0],[3,1]]
    },
].forEach(t => {
    console.log("Order of", t.numCourses, "courses depicted by", t.prerequisites,
                "->", findOrder(t.numCourses, t.prerequisites));
});

File: p210m/course_schedule2.py
from typing import List
from collections import defaultdict


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = defaultdict(list)
        for c, pre in prerequisites:
            graph[c].append(pre)

        ans = []
        visited = [False for _ in range(numCourses)]
        onpath = [False for _ in range(numCourses)]

        # return true if there is cycle in the graph
        def topologicalSort(v: int) -> bool:
            if visited[v]:
                return False
            visited[v], onpath[v] = True, True
            for i in graph[v]:
                if onpath[i] or topologicalSort(i):
                    return True
            onpath[v] = False
            ans.append(v)
            return False

        for v in range(numCourses):
            if not visited[v] and topologicalSort(v):
                return []
        return ans


# TESTS
for numCourses, prerequisites, expected in [
    (1, [], [0]),
    (2, [[0, 1]], [1, 0]),
    (
        2,
        [
            [0, 1],
            [1, 0],
        ],
        [],
    ),
    (
        4,
        [
            [0, 1],
            [1, 2],
            [3, 1],
        ],
        [2, 1, 0, 3],
    ),
    (
        4,
        [
            [1, 0],
            [2, 0],
            [3, 1],
            [3, 2],
        ],
        [0, 1, 2, 3],
    ),
]:
    sol = Solution()
    actual = sol.findOrder(numCourses, prerequisites)
    print("Order of", numCourses, "courses depicted by", prerequisites, "->", actual)
    assert actual == expected
